StyleCop.Analyzers/StyleCop.Analyzers/GeneratedCodeAnalysisExtensions.cs

```diff
@@ -0,0 +1,97 @@
+﻿// This file originally obtained from 
+// https://raw.githubusercontent.com/code-cracker/code-cracker/08c1a01337964924eeed12be8b14c8ce8ec6b626/src/Common/CodeCracker.Common/Extensions/GeneratedCodeAnalysisExtensions.cs
+// It is subject to the Apache License 2.0
+// This file has been modified since obtaining it from its original source.
+
+namespace StyleCop.Analyzers
+{
+    using System.Linq;
+    using System.Runtime.CompilerServices;
+    using System.Text.RegularExpressions;
+    using System.Threading;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.CodeAnalysis.CSharp.Syntax;
+    using Microsoft.CodeAnalysis.Diagnostics;
+
+    internal static class GeneratedCodeAnalysisExtensions
+    {
+        /// <summary>
+        /// A cache of the result of computing whether a document has an auto-generated header.
+        /// </summary>
+        /// <remarks>
+        /// This allows many analyzers that run on every token in the file to avoid checking
+        /// the same state in the document repeatedly.
+        /// </remarks>
+        private static readonly ConditionalWeakTable<SyntaxTree, StrongBox<bool?>> GeneratedHeaderPresentCheck = new ConditionalWeakTable<SyntaxTree, StrongBox<bool?>>();
+
+        internal static bool IsGenerated(this SyntaxNodeAnalysisContext context, CancellationToken cancellationToken) => (context.Node.SyntaxTree?.IsGenerated(cancellationToken) ?? false) || (context.Node?.IsGenerated() ?? false);
+
+        internal static bool IsGenerated(this SyntaxTreeAnalysisContext context, CancellationToken cancellationToken) => context.Tree?.IsGenerated(cancellationToken) ?? false;
+
+        internal static bool IsGenerated(this SymbolAnalysisContext context, CancellationToken cancellationToken)
+        {
+            if (context.Symbol == null) return false;
+            foreach (var syntaxReference in context.Symbol.DeclaringSyntaxReferences)
+            {
+                if (syntaxReference.SyntaxTree.IsGenerated(cancellationToken)) return true;
+                var root = syntaxReference.SyntaxTree.GetRoot(cancellationToken);
+                var node = root?.FindNode(syntaxReference.Span);
+                if (node.IsGenerated()) return true;
+            }
+            return false;
+        }
+
+        internal static bool IsGenerated(this SyntaxNode node) => node.HasAttributeOnAncestorOrSelf("DebuggerNonUserCode", "GeneratedCode");
```

Comment:
:exclamation: I disagree on `DebuggerNonUserCode`.

:question: Do we really need the `GeneratedCode` check? We are paying a heavy performance cost for this when most users would rely completely on file names and/or a comment in the header.